### PR TITLE
doc: add pre/postconditions and invariants to module keyvalue

### PR DESCRIPTION
**Pull request by lawli3t**
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: src/libs/elektra/keyvalue.c [hunk 1]
```diff
@@ -353,6 +362,10 @@ ssize_t keyGetString (const Key * key, char * returnedString, size_t maxSize)
  * regardless of the program's current encoding (if the iconv plugin is
  * available).
  *
+ * @pre @p newStringValue is a NULL terminated string
+ * @post Value of the Key is set to the UTF-8 encoded value of @p newStringValue
+ * @post Metadata keys "binary" / "origvalue" are cleared
```
Comment (markus2330) [inline src/libs/elektra/keyvalue.c:367]: origvalue is a hack with a wrong name (it should be `meta:/original/value` or similar), we shouldn't have this as part of our API.

See #3626 for more information about origvalue.

```suggestion
 * @post Metakey `meta:/binary` is cleared
```

File: src/libs/elektra/keyvalue.c [hunk 2]
```diff
@@ -480,6 +493,11 @@ ssize_t keyGetBinary (const Key * key, void * returnedBinary, size_t maxSize)
  * won't be changed. This will only happen in the successful case,
  * but not when -1 is returned.
  *
+ * @pre @p newBinary is not NULL and @p dataSize > 0
+ * @pre @p key is not read-only
+ * @post @p key's value set exactly to the data in @p newBinary
```
Comment (markus2330) [inline src/libs/elektra/keyvalue.c:498]: What if dataSize does not match the size of newBinary? I think we need one more precondition.

**Review by markus2330 — APPROVED** [review]
Very good design by contract :sparkling_heart:

**Comment by mpranj** [comment]
@lawli3t can you fix/address the open review points?

File: src/libs/elektra/keyvalue.c [hunk 3]
```diff
@@ -148,6 +148,9 @@ ksDel (ks);
 kdbClose (handle);
  * @endcode
  *
+ * @post returned pointer points to the stored internal value
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply src/libs/elektra/keyvalue.c:151]: not sure I see it 100%, what do you mean exactly?

**Comment by lawli3t** [comment]
> @lawli3t can you fix/address the open review points?

on it

File: src/libs/elektra/keyvalue.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpranj) [reply src/libs/elektra/keyvalue.c:151]: The key needs to contain data (i.e. key and key->data.v are not NULL)? Maybe that is meant

**Comment by markus2330** [comment]
Great work, thank you!

**Pull request merged**